Fix DataSourceUtils inconsistent exception handling

kevin0x90 opened the pull request:
Align IllegalStateException with SQLException handling since IllegalState is only handled internally and the public facing Exception remains unchanged in either way.

Closes gh-28258

snicoll commented on spring-jdbc/src/main/java/org/springframework/jdbc/datasource/DataSourceUtils.java:
```diff
@@ -151,14 +148,13 @@ public static Connection doGetConnection(DataSource dataSource) throws SQLExcept
 		Connection con = dataSource.getConnection();
 		if (con == null) {
-			throw new IllegalStateException("DataSource returned null from getConnection(): " + dataSource);
+			throw new SQLException("DataSource returned null from getConnection(): " + dataSource);
```
That change doesn't sound right to me. We shouldn't be throwing `SQLException` here.

kevin0x90 replied:
I did this change because the `CannotGetJdbcConnectionException` only accepts SQLException as cause parameter. 
So what do you think of relaxing the type constraint in `CannotGetJdbcConnectionException` a bit so that also IllegalStateException could be provided either via a constructor overload or a more relaxed type specified.

kevin0x90 replied:
So i would interpret this like the PR could be just closed.

kevin0x90 replied:
Ok i will adjust the code accordingly.

snicoll reviewed (commented):
Thanks for the PR but I don't think we should be changing the nested cause the way you suggest.

kevin0x90 committed:
Fix DataSourceUtils inconsistent exception handling

Align IllegalStateException with SQLException handling and propagate the original exception

Closes gh-28258

snicoll commented:
Good stuff, thanks @kevin0x90 and @Suthman.

kevin0x90 commented:
Thanks for accepting our small contribution @snicoll
